### PR TITLE
IAP header support

### PR DESCRIPTION
**Pull request by willwu0129**

### CODE REVIEW

File: src/api_manager/check_auth.cc [hunk 1]
```diff
@@ -44,6 +44,7 @@ namespace api_manager {
 namespace {
 const char kAccessTokenName[] = "access_token";
 const char kAuthHeader[] = "authorization";
+const char kAuthHeaderIAP[] = "X-Goog-Iap-Jwt-Assertion";
```
Comment (qiwzhang) [inline src/api_manager/check_auth.cc]: use all lower case
http header are case in-sensitive.

Comment (willwu0129) [inline src/api_manager/check_auth.cc]: Done.

File: src/api_manager/check_auth.cc [hunk 2]
```diff
@@ -170,6 +171,10 @@ void AuthChecker::Check() {
 void AuthChecker::GetAuthToken() {
   Request *r = context_->request();
   std::string auth_header;
+  if (r->FindHeader(KAuthHeaderIAP, &auth_header)) {
```
Comment (qiwzhang) [inline src/api_manager/check_auth.cc]: It value is just token?  or prefixed with "Bearer "?

Comment (qiwzhang) [inline src/api_manager/check_auth.cc]: Please add comment here to describe the value format

Comment (willwu0129) [inline src/api_manager/check_auth.cc]: Yes, the value is just token without prefix "Bearer". I will add a explanation.

Comment (qiwzhang) [inline src/api_manager/check_auth.cc]: second issue is order or precedence or both authorization and this one is set.  
I think we should take "authorization" first.  if not there,  then check this one

File: src/api_manager/check_auth.cc [hunk 3]
```diff
@@ -172,13 +172,13 @@ void AuthChecker::GetAuthToken() {
...
   if (!r->FindHeader(kAuthHeader, &auth_header)) {
+    // IAP header is of format "X-Goog-Iap-Jwt-Assertion": "eyJhbG...". No
+    // "Bearer" prefix is needed.
+    if (r->FindHeader(kAuthHeaderIAP, &auth_header)) {
+      auth_token_ = auth_header;
+      return;
+    }
     // When authorization header is missing, check query parameter.
     r->FindQuery(kAccessTokenName, &auth_token_);
     return;
```
Comment (qiwzhang) [inline src/api_manager/check_auth.cc]: How about after line 183?

Comment (willwu0129) [inline src/api_manager/check_auth.cc]: Okay:)

File: src/api_manager/check_auth.cc [hunk 4]
```diff
@@ -170,7 +171,14 @@ void AuthChecker::Check() {
...
   Request *r = context_->request();
   std::string auth_header;
+
   if (!r->FindHeader(kAuthHeader, &auth_header)) {
+    // IAP header is of format "X-Goog-Iap-Jwt-Assertion": "eyJhbG...". No
+    // "Bearer" prefix is needed.
+    if (r->FindHeader(kAuthHeaderIAP, &auth_header)) {
+      auth_token_ = auth_header;
+      return;
+    }
```
Comment (liminw) [inline src/api_manager/check_auth.cc]: Let's put the check for IAP header before checking "authorization" header. It's possible that IAP header and authorization header both exist (customer use authorization header for their own auth purpose). In that case, ESP will just check auth token in IAP header.

Comment (willwu0129) [inline src/api_manager/check_auth.cc]: Done.

**Pull request merged**